crypto/src/lib.rs

```diff
@@ -0,0 +1,117 @@
+    }
+}
+
+pub trait Hasher: Sized {
+    /// Adds a chunk to the running hash.
+    ///
+    /// # Arguments
+    ///
+    /// * `bytes` - Value to add to hash.
+    fn update(&mut self, bytes: &[u8]) -> Option<()>;
+
+    /// Finish a running hash operation and return the result.
+    ///
+    /// Once this function has been called, the object can no longer be used and
+    /// a new one must be created to hash more data.
+    ///
+    /// # Arguments
+    ///
+    /// * `digest` - Where the computed digest should be written.
+    fn finish<const SIZE: usize>(self) -> Option<[u8; SIZE]>;
+}
+
+pub trait Crypto {
+    type Cdi;
+    type Hasher: Hasher;
+
+    /// Fills the buffer with random values.
+    ///
+    /// # Arguments
+    ///
+    /// * `dst` - The buffer to be filled.
+    fn rand_bytes(dst: &mut [u8]) -> Option<()>;
+
+    /// Cryptographically hashes the given buffer.
+    ///
+    /// # Arguments
+    ///
+    /// * `profile` - Which profile is being used. This will tell the platform which algorithm to
+    ///   use.
+    /// * `bytes` - Value to be hashed.
+    /// * `digest` - Where the computed digest should be written.
+    fn hash<const SIZE: usize>(algs: AlgLen, bytes: &[u8]) -> Option<[u8; SIZE]> {
+        assert_eq!(SIZE, algs.size());
+        let mut hasher = Self::hash_initialize(algs)?;
+        hasher.update(bytes)?;
+        hasher.finish()
+    }
+
+    /// Initialize a running hash. Returns an object that will be able to complete the rest.
+    ///
+    /// Used for hashing multiple buffers that may not be in consecutive memory.
+    ///
+    /// # Arguments
+    ///
+    /// * `profile` - Which profile is being used. This will tell the platform which algorithm to
+    ///   use.
+    fn hash_initialize(algs: AlgLen) -> Option<Self::Hasher>;
+
+    /// Derive a CDI based on the current base CDI and measurements.
+    ///
+    /// # Arguments
+    ///
+    /// * `profile` - Which profile is being used. This will tell the platform
+    ///   which algorithm to use
+    /// * `measurement_digest` - A digest of the measurements which should be
+    ///   used for CDI derivation
+    /// * `info` - Caller-supplied info string to use in CDI derivation
+    fn derive_cdi(algs: AlgLen, measurement_digest: &[u8], info: &[u8]) -> Option<Self::Cdi>;
+
+    /// Derives an ECDSA keypair from `cdi` and returns the public key
+    ///
+    /// # Arguments
+    ///
+    /// * `profile` - Which profile is being used. This will tell the platform
+    ///    which algorithm to use
+    /// * `cdi` - CDI from which to derive the signing key
+    /// * `label` - Caller-supplied label to use in asymmetric key derivation
+    /// * `info` - Caller-supplied info string to use in asymmetric key derivation
+    ///
+    /// Returns a derived public key
+    fn derive_ecdsa_pub<const SIZE: usize>(
+        algs: AlgLen,
+        cdi: &Self::Cdi,
+        label: &[u8],
+        info: &[u8],
+    ) -> Option<([u8; SIZE], [u8; SIZE])>;
```

Comment:
I changed these to make it easier for the compiler to catch us using the wrong sized arrays.